### PR TITLE
[CGP] Consider arguments and ret values in `dupRetToEnableTailCallOpts`

### LINKED ISSUE #75455: Missed optimization: not tail calling memset
The code is (extracted from calloc implementation):
```
void* foo(void* x, long s) {
    if (__builtin_expect(x != nullptr, true))
      return __builtin_memset(x, 0, s);
    return x;
}
```

Clang generages:
```
foo(void*, long):                              # @foo(void*, long)
        pushq   %rbx
        movq    %rdi, %rbx
        testq   %rdi, %rdi
        je      .LBB0_2
        movq    %rsi, %rdx
        movq    %rbx, %rdi
        xorl    %esi, %esi
        callq   memset@PLT
.LBB0_2:
        movq    %rbx, %rax
        popq    %rbx
        retq
```
https://godbolt.org/z/YojTG44bs

This should be:
```
foo(void*, long):                              # @foo(void*, long)
        testq   %rdi, %rdi
        je      .LBB0_2
        movq    %rsi, %rdx
        xorl    %esi, %esi
        jmp     memset@PLT
.LBB0_2:
        retq
```

### PR DESCRIPTION
**Pull request by antoniofrighetto**
Hint further tail call optimization opportunities when the examined returned value is either the return value of a call instruction, or a function argument. Moreover, take into account the cases in which incoming values from phi-nodes, not directly tail call instructions, may still be simplified.

Fixes: https://github.com/llvm/llvm-project/issues/75455.

### CODE REVIEW

File: llvm/lib/CodeGen/CodeGenPrepare.cpp
Comment (dtcxzyw) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:2536]: Could you please move this line before the `dyn_cast<PHINode>` call?

Comment (antoniofrighetto) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp:2536]: May be worth it to anticipate before the `dyn_cast<BitCastInst>`?

**Comment by dtcxzyw** [comment]
@antoniofrighetto Could you please take a look at `llvm-test-suite/MultiSource/Benchmarks/Olden/health/health`? This patch seems to cause a significant regression (~22%) in this application, which is compiled with `-Os`.

**Comment by antoniofrighetto** [comment]
@dtcxzyw, thanks, restrained the folding to ret values of function calls and function arguments, and considering essentially only libc functions (whose first argument coincides with the return value).

**Comment by dtcxzyw** [comment]
> @dtcxzyw, thanks, restrained the folding to ret values of function calls and function arguments, and considering essentially only libc functions (whose first argument coincides with the return value).

The regression has been fixed. Thank you!

**Comment by antoniofrighetto** [comment]
Kind ping.

**Comment by dvyukov** [comment]
cc @david-xl @teresajohnson can you please find somebody to review this?
This is extracted from our production code, see #75455.

**Comment by dvyukov** [comment]
@antoniofrighetto do you have any intuition regarding how much code this can affect in real applications?

File: llvm/lib/CodeGen/CodeGenPrepare.cpp
Comment (david-xl) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:2641]: Is there a need to check which arg is fed to the phi and returned? The arg needs to be the real return value of the lowered function call?

**Comment by antoniofrighetto** [comment]
@dvyukov, hard to assess the impact of the change qualitatively. Looking for calls in tail position w/ ret value used over glibc (`grep -r -P 'return.*\b(strcpy|strncpy|memset|memcpy|memmove).*\(.*\).*;' .`) reveals 52 instances of tail call opportunities, which possibly doesn't seem a huge number (out of total # functions?), but they all seem within highly recurring functions (e.g., malloc, strdup – as already noted in the issue).

**Comment by dtcxzyw** [comment]
> @dvyukov, hard to assess the impact of the change qualitatively. Looking for calls in tail position w/ ret value used over glibc (`grep -r -P 'return.*\b(strcpy|strncpy|memset|memcpy|memmove).*\(.*\).*;' .`) reveals 52 instances of tail call opportunities, which possibly doesn't seem a huge number (out of total # functions?), but they all seem within highly recurring functions (e.g., malloc, strdup – as already noted in the issue).

I found 342 potential tail call opportunities in my [benchmark](https://github.com/dtcxzyw/llvm-opt-benchmark). But most of them are only associated with `llvm.memcpy/llvm.memset`.

File: llvm/lib/CodeGen/CodeGenPrepare.cpp
Comment (david-xl) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp:2641]: My question is if this case https://godbolt.org/z/jndea9van is properly detected and skipped.

Comment (antoniofrighetto) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp:2641]: Oh sure, thanks, added an extra check skipping such cases and this one (https://godbolt.org/z/3fce9qrvz) as well – we need the argument fed to the phi to be the first one.

Comment (antoniofrighetto) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp:2641]: Reviewed test too.

**Comment by antoniofrighetto** [comment]
Kind ping (cc/ @david-xl).

**Comment by dtcxzyw** [comment]
Ping.

**Comment by antoniofrighetto** [comment]
Updated this to explicitly looking for candidate intrinsics / `strcpy`, as I don't think only looking for first argument was robust enough to avoid phi folding to ret in all cases (now evaluating it on the compile-time tracker for regressions).

File: llvm/test/CodeGen/X86/tailcall-cgp-dup.ll
Comment (david-xl) [inline llvm/test/CodeGen/X86/tailcall-cgp-dup.ll:238]: Perhaps add a test case with diamond shaped control flow. In one arm, add a positive test and in the other arm, add a negative test where tailcall is not legal.

Comment (antoniofrighetto) [reply llvm/test/CodeGen/X86/tailcall-cgp-dup.ll]: Added https://clang.godbolt.org/z/a3rns4883, thanks.

**Comment by antoniofrighetto** [comment]
Ping.

**Comment by david-xl** [comment]
Can you provide any performance numbers (real applications, benchmark suite) etc?

**Comment by dvyukov** [comment]
> Can you provide any performance numbers (real applications, benchmark suite) etc?

What will depend on these numbers? Do you mean that we don't merge this if it's not possible to demonstrate the effect? Obtaining relativistic performance data for smaller improvements is usually very expensive, and not possible to do an improvement less than 0.1%.

**Comment by david-xl** [comment]
> > Can you provide any performance numbers (real applications, benchmark suite) etc?
> 
> What will depend on these numbers? Do you mean that we don't merge this if it's not possible to demonstrate the effect? Obtaining relativistic performance data for smaller improvements is usually very expensive, and not possible to do an improvement less than 0.1%.

It is the other way around. Having some positive numbers will help speed up the approval. There are always trade-offs between runtime performance and compile time overhead (probably not a concern for this case).

There is also chance of trigging unexpected regressions. Having some test coverage gives people more confidence.

Having said that, this patch LGTM.

File: llvm/test/CodeGen/X86/tailcall-cgp-dup.ll
Comment (nikic) [inline llvm/test/CodeGen/X86/tailcall-cgp-dup.ll:1]: Mark the functions as nounwind instead.

File: llvm/lib/CodeGen/CodeGenPrepare.cpp
Comment (nikic) [inline llvm/lib/CodeGen/CodeGenPrepare.cpp:2675]: Why are these extra checks only needed in the PN branch, but not the !PN branch?

Comment (antoniofrighetto) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp:2679]: It turns out we need them in the !PN branch too when handling a Arg or CallInst, added more test coverage, thanks for taking a look.

File: llvm/lib/CodeGen/CodeGenPrepare.cpp
Comment (antoniofrighetto) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp]: Anticipated `isIntrinsicOrLFToBeTailCalled` call so that we are guaranteed the first argument operand always exists.

File: llvm/lib/CodeGen/CodeGenPrepare.cpp
Comment (antoniofrighetto) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp]: Right. I thought checking that `V` exists would suffice, but it’s not the case. Checking directly that `V` is the first argument, adding another test case.

File: llvm/lib/CodeGen/CodeGenPrepare.cpp
Comment (antoniofrighetto) [reply llvm/lib/CodeGen/CodeGenPrepare.cpp]: Yes, now looks reasonable to drop the whole check after further restricting the !PN path.

**Review by nikic — APPROVED** [review]
LGTM

**Comment by antoniofrighetto** [comment]
Thanks for reviewing this! @david-xl, @nikic

**Pull request merged**
